kernel/comps/virtio/src/device/network/device.rs

```diff
@@ -170,29 +191,90 @@ fn queue_to_network_error(err: QueueError) -> VirtioNetError {
     match err {
         QueueError::NotReady => VirtioNetError::NotReady,
         QueueError::WrongToken => VirtioNetError::WrongToken,
+        QueueError::BufferTooSmall => VirtioNetError::Busy,
         _ => VirtioNetError::Unknown,
     }
 }
 
+fn init_caps(features: &NetworkFeatures, config: &VirtioNetConfig) -> DeviceCapabilities {
+    let mut caps = DeviceCapabilities::default();
+
+    caps.max_burst_size = None;
+    caps.medium = Medium::Ethernet;
+
+    if features.contains(NetworkFeatures::VIRTIO_NET_F_MTU) {
+        // If `VIRTIO_NET_F_MTU` is negotiated, the MTU is decided by the device.
+        caps.max_transmission_unit = config.mtu as usize;
+    } else {
+        // Linux does not has a specified MTU for virtio-net.
+        // Linux only requires the MTU between 68-65535.
+        // The exact value can be decided dynamically.
+        // <https://elixir.bootlin.com/linux/v6.0.9/source/drivers/net/virtio_net.c#L3660>.
+
+        if features.contains(NetworkFeatures::VIRTIO_NET_F_GUEST_TSO4)
+            || features.contains(NetworkFeatures::VIRTIO_NET_F_GUEST_TSO6)
+            || features.contains(NetworkFeatures::VIRTIO_NET_F_GUEST_UFO)
+        {
+            // If any of the above features is set, the max incoming packets is 65550 bytes
+            caps.max_transmission_unit = 65600; // 65550 align up by 100
+        } else {
+            // Else, the incoming packets is 1514 bytes.
+            caps.max_transmission_unit = 1600; // 1514 align up by 100
+        }
+    }
+
+    // For transmitting: If `VIRTIO_NET_F_CSUM` is _not_ negotiated,
+    // the driver must supply a fully checksummed packet to the device.
+    // For receiving: If `VIRTIO_NET_F_GUEST_CSUM` is negotiated,
+    // the device has validated the packet checsum.
```

Comment:
typo: `checsum` -> `checksum`, but more importantly:

> 5.1.6.2 Packet Transmission
[ .. ]
If the driver negotiated VIRTIO_NET_F_CSUM, it can skip checksumming the packet:
• flags has the VIRTIO_NET_HDR_F_NEEDS_CSUM set,
• csum_start is set to the offset within the packet to begin checksumming, and
• csum_offset indicates how many bytes after the csum_start the new (16 bit ones’ complement)
checksum is placed by the device.
• The TCP checksum field in the packet is set to the sum of the TCP pseudo header, so that replacing
it by the ones’ complement checksum of the TCP header and body will give the correct result.
>
> Note: For example, consider a partially checksummed TCP (IPv4) packet. It will have a 14 byte ethernet header and 20 byte IP header followed by the TCP header (with the TCP checksum field 16 bytes into that header). csum_start will be 14+20 = 34 (the TCP checksum includes the header), and csum_offset will be 16.

Well, we do none of these matters above.

> 5.1.6.4 Processing of Incoming Packets
>
> If the VIRTIO_NET_F_GUEST_CSUM feature was negotiated, the VIRTIO_NET_HDR_F_DATA_VALID
bit in flags can be set: if so, device has validated the packet checksum

We never check this flag either.

This effectively means that checksum capabilities need to be handled on a per-packet logic. We don't have the ability to do that now, so there should just be a FIXME and take the most conservative choice.

Reply:
Yes, you are right. We actually didn't support these two features so we can't offload the checksum to device. We need to check checksum for both receiving and transmission.